### ISSUE TITLE
H/2 stress many failure in content posting methods 

### ISSUE THREAD

**Issue by ManickaP** [dotnet/runtime#55261]
Reproduces in and out of Docker. Started happening between 6.0 Preview 4 and **6.0 Preview 6** (on 4 it ran, on 6 started failing) - tested locally.
Error:
```
System.Exception: Expected status code OK, got InternalServerError
   at HttpStress.ClientOperations.ValidateStatusCode(HttpResponseMessage m, HttpStatusCode expectedStatus) in /home/manicka/repositories/runtime/src/libraries/System.Net.Http/tests/StressTests/HttpStress/ClientOperations.cs:line 487
   at HttpStress.ClientOperations.<>c.<<get_Operations>b__1_12>d.MoveNext() in /home/manicka/repositories/runtime/src/libraries/System.Net.Http/tests/StressTests/HttpStress/ClientOperations.cs:line 451
--- End of stack trace from previous location ---
   at HttpStress.StressClient.<>c__DisplayClass17_0.<<StartCore>g__RunWorker|0>d.MoveNext() in /home/manicka/repositories/runtime/src/libraries/System.Net.Http/tests/StressTests/HttpStress/StressClient.cs:line 204
```

Potentially kestrel error, cc: @JamesNK @Tratcher @halter73 

Discovered in https://github.com/dotnet/runtime/pull/55098

**Comment by ManickaP** [comment]
Some more digging:
kestrel log:
```
[15:25:43 ERR] Connection id "0HMA19AMUPRVO", Request id "0HMA19AMUPRVO:00003227": An unhandled exception was thrown by the application.
System.IO.IOException: The client reset the request stream.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestPipeReader.ReadAsync(CancellationToken cancellationToken) in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000db2+0x7
   at System.IO.Pipelines.PipeReader.CopyToAsyncCore[TStream](TStream destination, Func`4 writeAsync, CancellationToken cancellationToken) in System.IO.Pipelines.dll:token 0x60000b8+0x236
   at HttpStress.StressServer.<>c.<<MapRoutes>b__6_6>d.MoveNext() in /home/manicka/repositories/runtime/src/libraries/System.Net.Http/tests/StressTests/HttpStress/StressServer.cs:line 254
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger) in Microsoft.AspNetCore.Routing.dll:token 0x60000ab+0xa9
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application) in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000d28+0x54c
```

Happens with slow content sending operation much more:
NO cancellation:
```
[7/7/2021 3:32:53 PM] Total: 7,076 Runtime: 00:00:06
	 0: GET                       Success: 469	Canceled: 0	Fail: 0
	 1: GET Partial               Success: 470	Canceled: 0	Fail: 0
	 2: GET Headers               Success: 471	Canceled: 0	Fail: 0
	 3: GET Parameters            Success: 472	Canceled: 0	Fail: 0
	 4: GET Aborted               Success: 472	Canceled: 0	Fail: 0
	 5: POST                      Success: 456	Canceled: 0	Fail: 16
	 6: POST Multipart Data       Success: 452	Canceled: 0	Fail: 21
	 7: POST Duplex               Success: 465	Canceled: 0	Fail: 9
	 8: POST Duplex Slow          Success: 459	Canceled: 2	Fail: 13
	 9: POST Duplex Dispose       Success: 456	Canceled: 0	Fail: 16
	10: POST ExpectContinue       Success: 458	Canceled: 0	Fail: 14
	11: HEAD                      Success: 472	Canceled: 0	Fail: 0
	12: PUT                       Success: 450	Canceled: 0	Fail: 22
	13: PUT Slow                  Success: 453	Canceled: 3	Fail: 16
	14: GET Slow                  Success: 468	Canceled: 1	Fail: 0
	    TOTAL                     Success: 6,943	Canceled: 6	Fail: 127

```
Cancellation:
```
[7/7/2021 3:32:18 PM] Total: 6,540 Runtime: 00:00:06
	 0: GET                       Success: 407	Canceled: 27	Fail: 0
	 1: GET Partial               Success: 403	Canceled: 32	Fail: 0
	 2: GET Headers               Success: 398	Canceled: 38	Fail: 0
	 3: GET Parameters            Success: 406	Canceled: 30	Fail: 0
	 4: GET Aborted               Success: 404	Canceled: 32	Fail: 0
	 5: POST                      Success: 409	Canceled: 20	Fail: 8
	 6: POST Multipart Data       Success: 413	Canceled: 20	Fail: 5
	 7: POST Duplex               Success: 397	Canceled: 33	Fail: 8
	 8: POST Duplex Slow          Success: 405	Canceled: 28	Fail: 5
	 9: POST Duplex Dispose       Success: 403	Canceled: 28	Fail: 5
	10: POST ExpectContinue       Success: 391	Canceled: 37	Fail: 8
	11: HEAD                      Success: 412	Canceled: 24	Fail: 0
	12: PUT                       Success: 404	Canceled: 25	Fail: 7
	13: PUT Slow                  Success: 402	Canceled: 29	Fail: 3
	14: GET Slow                  Success: 405	Canceled: 29	Fail: 0
	    TOTAL                     Success: 6,059	Canceled: 432	Fail: 49
```

**Comment by karelz** [comment]
@JamesNK @Tratcher @halter73 is it something you could help us with please? It seems like Kestrel regression and it blocks our stress runs :(
Should we create issue in aspnet repo?

**Comment by halter73** [comment]
> System.IO.IOException: The client reset the request stream.

@karelz What makes this seem like a Kestrel regression? The above exception is only thrown if Kestrel received a `RST_STREAM` frame from the client.

> System.Exception: Expected status code OK, got InternalServerError

I'm assuming this means the client received a 500 response. This could be caused by an uncaught exception on the server, but not an exception caused by `RST_STREAM` since that closes the stream.

Can you collect more verbose server logs and/or collect a network trace?

**Comment by alnikola** [comment]
Client logs show that these HTTP/2 POST tests are failing only on Windows and only with timeout in one of the following 2 places:
1. [Http2Stream.WaitForDataAsync](https://github.com/dotnet/runtime/blob/1d8ad03abb72fe79c44030f2806a0109a7909d25/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L1375)
2. [Http2Stream.WaitFor100ContinueAsync](https://github.com/dotnet/runtime/blob/a2f2a629328a136c2cd18f2cd729e3854de2aeb4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L327)

The timeout exception message is
>System.Threading.Tasks.TaskCanceledException: The request was canceled due to the configured HttpClient.Timeout of 60 seconds elapsing.

**Comment by alnikola** [comment]
My first guess would be that the server hasn't responded in the given time.

**Comment by alnikola** [comment]
[My comment above](https://github.com/dotnet/runtime/issues/55261#issuecomment-884303936) above might not be related to this issue because here the server responds with 500 to client whereas I observed timeouts. I'll continue the investigation.

**Comment by alnikola** [comment]
It seems the issue is caused by the client sending EndOfStream without a RST_STREAM when it has set the Content-Length and the server is still reading the request body when EndOfStream arrives. It's a protocol violation so the server throws the following exception and responds with 500.

```
[DBG] Connection id "0HMAGA52UA8E0": HTTP/2 stream error.
Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2StreamErrorException: HTTP/2 stream ID 695 error (PROTOCOL_ERROR): Less data received than specified in the Content-Length header.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Stream.OnEndStreamReceived() in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000da7+0x39
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Stream.OnDataAsync(Http2Frame dataFrame, ReadOnlySequence`1& payload) in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000da6+0x141
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessDataFrameAsync(ReadOnlySequence`1& payload) in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000cff+0x12b
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessFrameAsync[TContext](IHttpApplication`1 application, ReadOnlySequence`1& payload) in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000cfe+0xc0
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessRequestsAsync[TContext](IHttpApplication`1 application) in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000cfa+0x9df
```

**Comment by alnikola** [comment]
It might indicate there is a bug with how [Http2Stream sends EndOfStream in SendRequestBodyAsync](https://github.com/dotnet/runtime/blob/0c200955af77bf17cd05488db96d3cdb3716a16c/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L288).

**Comment by Tratcher** [comment]
https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.6
>    A request or response that includes a payload body can include a
   content-length header field.  A request or response is also malformed
   if the value of a content-length header field does not equal the sum
   of the DATA frame payload lengths that form the body.

**Comment by adityamandaleeka** [comment]
Can someone please try to repro this failure on .NET 5? That would help us understand if this is a regression.

**Comment by antonfirsov** [comment]
This seemed to cause massive failures in the CI `stress-http` runs in the first week of July ([see here](https://dev.azure.com/dnceng/public/_build/results?buildId=1226831&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c56d998-158d-5232-283c-5140104799fc)) but the CI failures disappeared after 7/8.

**Comment by CarnaViire** [comment]
> This seemed to cause massive failures in the CI stress-http runs in the first week of July (see here) but the CI failures disappeared after 7/8.

AFAIK it started after updating to latest Kestrel version in https://github.com/dotnet/runtime/pull/55098 because we needed it for HTTP/3, but in the same PR we put "active issue" on the failures not to pollute runs. Most possibly the massive failures you've seen were for that PR only, not for main?

**Comment by karelz** [comment]
The test were disabled on 7/8, so that checks out ;)

@alnikola will you be able to run your repro easily against earlier version of Kestrel (either 5.0, or earlier 6.0 Preview)?

**Comment by alnikola** [comment]
I will try to run stress tests on 6.0 Preview 4 (it's reported to be healthy) or .NET 5.

**Comment by alnikola** [comment]
@adityamandaleeka I ran these POST tests on .NET Core: 5.0.8, ASP.NET Core: 5.0.8 and the completed successfully. I have not got any errors.

**Comment by karelz** [comment]
@alnikola would it be possible to use latest .NET Core libraries for client side, while using old Kestrel? That would tell us if our current client has a bug or if it is Kestrel.

**Comment by adityamandaleeka** [comment]
Thanks @alnikola, that's very useful info. I'll second @karelz's question too :) .

**Comment by alnikola** [comment]
Strangely enough, but I cannot anymore reproduce this issue on my local machine on any .NET versions. Even the `6.0.0-preview.6.21352.12` where it failed initially, now works. It's not clear what could change in the last few days. I didn't install any new SDKs or runtime versions since then.

**Comment by alnikola** [comment]
>would it be possible to use latest .NET Core libraries for client side, while using old Kestrel? That would tell us if our current client has a bug or if it is Kestrel.

I tested with `HttpClient [6.0.0-preview.6.21352.12]` and `Asp.Net Core [5.0.8]`. All runs completed successfully. No failures.

**Comment by adityamandaleeka** [comment]
Thank you for confirming, @alnikola. We'll prioritize investigating this regression.

**Comment by adityamandaleeka** [comment]
@alnikola Can you please share instructions for how to run the stress test that reproduces this?

**Comment by alnikola** [comment]
> Can you please share instructions for how to run the stress test that reproduces this?

First, globally install the required SDK and runtime versions because stress tests use the globally installed .NET.

Then, in `runtime` repo:
1. Go to the folder `src/libraries/System.Net.Http/tests/StressTests/HttpStress`
2. Open 2 console windows in this folder
3. In the first window run the command to start server `dotnet run -- -runMode server -serverUri https://+:5001`
4. In the second window, run the command to start client `dotnet run -- -runMode client -serverUri https://localhost:5001 -ops 5 6 7 8 9 10 -maxContentLength 10000`
5. Watch the client's output and wait for exceptions to be thrown and logged into the console

**Comment by ManickaP** [comment]
Depends on https://github.com/dotnet/aspnetcore/issues/34768

**Comment by karelz** [comment]
Reopening to track backport to 6.0.

**Comment by karelz** [comment]
Fixed in 6.0 RC2 in PR #58541.
Fixed in 7.0 in PR #58442.
